File: lib/activity.py
#!/usr/bin/env python3
"""
Activity Logging Module

Tracks user activities with details and timestamps.
Stores activities in JSON format for easy parsing and querying.
"""

import json
import os
from datetime import datetime
from pathlib import Path
from typing import Any


class ActivityLogger:
    """Logger for user activities."""

    ACTIVITY_TYPES = {
        "login": "User logged in",
        "logout": "User logged out",
        "config_change": "Configuration changed",
        "input_change": "Input value changed",
        "save": "Data saved",
        "load": "Data loaded",
        "error": "Error occurred",
        "warning": "Warning issued",
        "info": "Informational event",
    }

    def __init__(self, log_dir: str | None = None):
        """Initialize the activity logger.

        Args:
            log_dir: Directory to store activity logs. Defaults to /data/.openclaw/activity
        """
        self.log_dir = Path(
            log_dir or os.environ.get("ACTIVITY_LOG_DIR", "/data/.openclaw/activity")
        )
        self.log_dir.mkdir(parents=True, exist_ok=True)
        self.current_log_file = (
            self.log_dir / f"activities_{datetime.now().strftime('%Y-%m-%d')}.jsonl"
        )
        self.enabled = os.environ.get("ACTIVITY_LOG_ENABLED", "true").lower() == "true"

    def log(
        self,
        user: str,
        activity_type: str,
        details: dict[str, Any] | None = None,
        source: str = "system",
    ) -> dict[str, Any]:
        """Log a user activity.

        Args:
            user: The user who performed the activity
            activity_type: Type of activity (login, save, config_change, etc.)
            details: Additional details about the activity
            source: Source of the activity (web, cli, api, system)

        Returns:
            The logged activity entry
        """
        if not self.enabled:
            return {}

        activity = {
            "timestamp": datetime.utcnow().isoformat() + "Z",
            "user": user,
            "activity": activity_type,
            "description": self.ACTIVITY_TYPES.get(activity_type, activity_type),
            "source": source,
            "details": details or {},
        }

        # Append to JSONL file
        with open(self.current_log_file, "a") as f:
            f.write(json.dumps(activity) + "\n")

        return activity

    def get_activities(
        self,
        user: str | None = None,
        activity_type: str | None = None,
        start_time: str | None = None,
        end_time: str | None = None,
        limit: int = 100,
        offset: int = 0,
    ) -> list[dict[str, Any]]:
        """Retrieve activities with optional filtering.

        Args:
            user: Filter by username
            activity_type: Filter by activity type
            start_time: Filter activities after this time (ISO format)
            end_time: Filter activities before this time (ISO format)
            limit: Maximum number of activities to return
            offset: Number of activities to skip

        Returns:
            List of activity entries
        """
        activities = []

        # Read all log files
        log_files = sorted(self.log_dir.glob("activities_*.jsonl"), reverse=True)

        for log_file in log_files:
            if not log_file.exists():
                continue

            with open(log_file) as f:
                for line in f:
                    line = line.strip()
                    if not line:
                        continue

                    try:
                        activity = json.loads(line)
                    except json.JSONDecodeError:
                        continue

                    # Apply filters
                    if user and activity.get("user") != user:
                        continue
                    if activity_type and activity.get("activity") != activity_type:
                        continue
                    if start_time and activity.get("timestamp", "") < start_time:
                        continue
                    if end_time and activity.get("timestamp", "") > end_time:
                        continue

                    activities.append(activity)

        # Sort by timestamp (newest first) and apply pagination
        activities.sort(key=lambda x: x.get("timestamp", ""), reverse=True)
        return activities[offset : offset + limit]

    def get_activity_stats(self, days: int = 7) -> dict[str, Any]:
        """Get activity statistics for the specified number of days.

        Args:
            days: Number of days to analyze

        Returns:
            Dictionary with activity statistics
        """
        from collections import Counter

        activities = self.get_activities(limit=10000)

        # Filter to specified days
        cutoff = datetime.utcnow().timestamp() - (days * 24 * 60 * 60)
        recent_activities = [
            a
            for a in activities
            if datetime.fromisoformat(a["timestamp"].replace("Z", "+00:00")).timestamp() > cutoff
        ]

        stats = {
            "total_activities": len(recent_activities),
            "unique_users": len(set(a["user"] for a in recent_activities)),
            "activity_types": dict(Counter(a["activity"] for a in recent_activities)),
            "top_users": dict(Counter(a["user"] for a in recent_activities).most_common(10)),
            "period_days": days,
        }

        return stats


def get_logger() -> ActivityLogger:
    """Get the singleton activity logger instance."""
    if not hasattr(get_logger, "_instance"):
        get_logger._instance = ActivityLogger()
    return get_logger._instance


# Convenience functions for common activities
def log_login(user: str, source: str = "web", details: dict | None = None) -> dict:
    """Log a user login."""
    return get_logger().log(user, "login", details, source)


def log_logout(user: str, source: str = "web") -> dict:
    """Log a user logout."""
    return get_logger().log(user, "logout", {}, source)


def log_config_change(user: str, changes: dict, source: str = "web") -> dict:
    """Log configuration changes."""
    return get_logger().log(user, "config_change", {"changes": changes}, source)


def log_input_change(
    user: str, field: str, old_value: Any, new_value: Any, source: str = "web"
) -> dict:
    """Log an input field change."""
    return get_logger().log(
        user,
        "input_change",
        {"field": field, "old_value": str(old_value), "new_value": str(new_value)},
        source,
    )


def log_save(user: str, item: str, source: str = "web") -> dict:
    """Log a save operation."""
    return get_logger().log(user, "save", {"item": item}, source)


def log_load(user: str, item: str, source: str = "web") -> dict:
    """Log a load operation."""
    return get_logger().log(user, "load", {"item": item}, source)


def log_error(user: str, error: str, details: dict | None = None, source: str = "system") -> dict:
    """Log an error."""
    return get_logger().log(user, "error", {"error": error, **(details or {})}, source)


def log_warning(
    user: str, warning: str, details: dict | None = None, source: str = "system"
) -> dict:
    """Log a warning."""
    return get_logger().log(user, "warning", {"warning": warning, **(details or {})}, source)


def log_info(user: str, message: str, details: dict | None = None, source: str = "system") -> dict:
    """Log an informational message."""
    return get_logger().log(user, "info", {"message": message, **(details or {})}, source)


if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="Activity Logger CLI")
    parser.add_argument("command", choices=["list", "stats", "log"])
    parser.add_argument("--user", help="Filter by user")
    parser.add_argument("--type", help="Filter by activity type")
    parser.add_argument("--limit", type=int, default=100, help="Limit results")
    parser.add_argument("--days", type=int, default=7, help="Days for stats")

    args = parser.parse_args()

    logger = get_logger()

    if args.command == "list":
        activities = logger.get_activities(
            user=args.user,
            activity_type=args.type,
            limit=args.limit,
        )
        print(json.dumps(activities, indent=2))
    elif args.command == "stats":
        stats = logger.get_activity_stats(days=args.days)
        print(json.dumps(stats, indent=2))
    elif args.command == "log":
        print(f"Activity log directory: {logger.log_dir}")
        print(f"Current log file: {logger.current_log_file}")
        print(f"Logging enabled: {logger.enabled}")

File: lib/activity.js
/**
 * Activity Logging Module (Node.js)
 * 
 * Tracks user activities with details and timestamps.
 * Stores activities in JSON format for easy parsing and querying.
 */

const fs = require('fs');
const path = require('path');

const ACTIVITY_TYPES = {
  login: 'User logged in',
  logout: 'User logged out',
  config_change: 'Configuration changed',
  input_change: 'Input value changed',
  save: 'Data saved',
  load: 'Data loaded',
  error: 'Error occurred',
  warning: 'Warning issued',
  info: 'Informational event',
};

class ActivityLogger {
  constructor(logDir = null) {
    this.logDir = logDir || process.env.ACTIVITY_LOG_DIR || '/data/.openclaw/activity';
    
    // Ensure directory exists
    if (!fs.existsSync(this.logDir)) {
      fs.mkdirSync(this.logDir, { recursive: true });
    }
    
    const date = new Date().toISOString().split('T')[0];
    this.currentLogFile = path.join(this.logDir, `activities_${date}.jsonl`);
    this.enabled = (process.env.ACTIVITY_LOG_ENABLED || 'true').toLowerCase() === 'true';
  }

  /**
   * Log a user activity
   * @param {string} user - The user who performed the activity
   * @param {string} activityType - Type of activity
   * @param {object} details - Additional details
   * @param {string} source - Source of the activity (web, cli, api, system)
   * @returns {object} The logged activity entry
   */
  log(user, activityType, details = {}, source = 'system') {
    if (!this.enabled) {
      return {};
    }

    const activity = {
      timestamp: new Date().toISOString(),
      user,
      activity: activityType,
      description: ACTIVITY_TYPES[activityType] || activityType,
      source,
      details,
    };

    // Append to JSONL file
    fs.appendFileSync(this.currentLogFile, JSON.stringify(activity) + '\n');

    return activity;
  }

  /**
   * Retrieve activities with optional filtering
   * @param {object} filters - Filter options
   * @returns {Array} List of activity entries
   */
  getActivities(filters = {}) {
    const { user, activityType, startTime, endTime, limit = 100, offset = 0 } = filters;
    const activities = [];

    // Read all log files
    const files = fs.readdirSync(this.logDir)
      .filter(f => f.startsWith('activities_') && f.endsWith('.jsonl'))
      .sort()
      .reverse();

    for (const file of files) {
      const filePath = path.join(this.logDir, file);
      const content = fs.readFileSync(filePath, 'utf8');
      const lines = content.split('\n').filter(line => line.trim());

      for (const line of lines) {
        try {
          const activity = JSON.parse(line);

          // Apply filters
          if (user && activity.user !== user) continue;
          if (activityType && activity.activity !== activityType) continue;
          if (startTime && activity.timestamp < startTime) continue;
          if (endTime && activity.timestamp > endTime) continue;

          activities.push(activity);
        } catch (e) {
          // Skip invalid JSON lines
        }
      }
    }

    // Sort by timestamp (newest first) and apply pagination
    activities.sort((a, b) => new Date(b.timestamp) - new Date(a.timestamp));
    return activities.slice(offset, offset + limit);
  }

  /**
   * Get activity statistics
   * @param {number} days - Number of days to analyze
   * @returns {object} Activity statistics
   */
  getActivityStats(days = 7) {
    const activities = this.getActivities({ limit: 10000 });
    
    const cutoff = Date.now() - (days * 24 * 60 * 60 * 1000);
    const recentActivities = activities.filter(a => 
      new Date(a.timestamp).getTime() > cutoff
    );

    const activityTypes = {};
    const users = {};

    for (const activity of recentActivities) {
      activityTypes[activity.activity] = (activityTypes[activity.activity] || 0) + 1;
      users[activity.user] = (users[activity.user] || 0) + 1;
    }

    // Get top activity type
    const topActivity = Object.entries(activityTypes)
      .sort((a, b) => b[1] - a[1])[0];

    return {
      total_activities: recentActivities.length,
      unique_users: Object.keys(users).length,
      activity_types: activityTypes,
      top_users: Object.entries(users)
        .sort((a, b) => b[1] - a[1])
        .slice(0, 10)
        .reduce((obj, [key, val]) => { obj[key] = val; return obj; }, {}),
      period_days: days,
    };
  }
}

// Singleton instance
let loggerInstance = null;

function getLogger() {
  if (!loggerInstance) {
    loggerInstance = new ActivityLogger();
  }
  return loggerInstance;
}

// Convenience functions
function logLogin(user, source = 'web', details = {}) {
  return getLogger().log(user, 'login', details, source);
}

function logLogout(user, source = 'web') {
  return getLogger().log(user, 'logout', {}, source);
}

function logConfigChange(user, changes, source = 'web') {
  return getLogger().log(user, 'config_change', { changes }, source);
}

function logInputChange(user, field, oldValue, newValue, source = 'web') {
  return getLogger().log(user, 'input_change', { 
    field, 
    old_value: String(oldValue), 
    new_value: String(newValue) 
  }, source);
}

function logSave(user, item, source = 'web') {
  return getLogger().log(user, 'save', { item }, source);
}

function logLoad(user, item, source = 'web') {
  return getLogger().log(user, 'load', { item }, source);
}

function logError(user, error, details = {}, source = 'system') {
  return getLogger().log(user, 'error', { error, ...details }, source);
}

function logWarning(user, warning, details = {}, source = 'system') {
  return getLogger().log(user, 'warning', { warning, ...details }, source);
}

function logInfo(user, message, details = {}, source = 'system') {
  return getLogger().log(user, 'info', { message, ...details }, source);
}

module.exports = {
  ActivityLogger,
  getLogger,
  logLogin,
  logLogout,
  logConfigChange,
  logInputChange,
  logSave,
  logLoad,
  logError,
  logWarning,
  logInfo,
  ACTIVITY_TYPES,
};
